### ISSUE TITLE
Skydive integration with NSM

### ISSUE THREAD

**Issue by edwarnicke** [ligato/networkservicemesh#502]
#490 is being worked by @haiodo to expose the [MonitorCrossConnect](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/apis/crossconnect/crossconnect.proto#L33) API.  This API is intended to be usable to feed [skydive.network](skydive.network).  

It should be possible to start working on the skydive integration in parallel because there is a library for creating a new [generic MonitorCrossConnect server](https://github.com/ligato/networkservicemesh/blob/master/controlplane/pkg/monitor_crossconnect_server/monitor_crossconnect_server.go#L39)

Basically, you can write a very very simple 'mock' MonitorCrossConnect server to start working against roughly like:
```go
ln, err := net.Listen("tcp", "0.0.0.0:9999")
if err != nil {
	logrus.Fatalf("Error listening on socket %s: %s ", dataplaneSocket, err)
}

monitor := monitor_crossconnect_server.NewMonitorCrossConnectServer()
monitor.UpdateCrossConnect(xcon1)
monitor.UpdateCrossConnect(xcon2)
monitor.DeleteCrossConnect(xcon1)

server := grpc.NewServer()
crossconnect.RegisterMonitorCrossConnectServer(server, monitor)
go server.Serve(ln)
```

Note, the above is the substantive part, you probably want to dress it up with:

```go
// Capture signals to cleanup before exiting
c := make(chan os.Signal, 1)
signal.Notify(c,
	syscall.SIGHUP,
	syscall.SIGINT,
	syscall.SIGTERM,
	syscall.SIGQUIT)
ln, err := net.Listen("tcp", "0.0.0.0:9999")
if err != nil {
	logrus.Fatalf("Error listening on socket %s: %s ", dataplaneSocket, err)
}

monitor := monitor_crossconnect_server.NewMonitorCrossConnectServer()
monitor.UpdateCrossConnect(xcon1)
monitor.UpdateCrossConnect(xcon2)
monitor.DeleteCrossConnect(xcon1)

server := grpc.NewServer()
crossconnect.RegisterMonitorCrossConnectServer(server, monitor)
go server.Serve(ln)

select {
    case <-c:
	logrus.Info("Closing Dataplane Registration")
	server.Stop()
}
```

to manage to have a server that hangs out until killed

**Comment by edwarnicke** [comment]
These [slides](https://docs.google.com/presentation/d/139fFJgtpsj0DcMOw24qiknEG9hk-1s1OSdzWCrb9EYk/edit#slide=id.g48255483fc_0_0) that were drawn while chatting with @yf-yang may also be helpful to others:

**Comment by yf-yang** [comment]
@matrohon 
Status Update:
I rearranged server/client into examples/skydive/[server|client] respectively, .monitor.mk is also changed accordingly.
In Client, I removed ctx.withBlock, so that it will raise Error when server is not available. I also added some comments, so that we can make this part better when move the client into SD probe.
In Server, I added fake CrossConnect generation mechanisms. I'm not sure how to fill some of the data fields, so feel free to change the generation process if the values are not accurate or the names are confusing.
I also did some trivial test to make sure the communication mechanism between Client and Server shown in the picture above is working. Now the client could receive values, so I think the next step is unpacking values from it.

All these updates are in branch `fake-monitors` in https://github.com/yf-yang/networkservicemesh, I've added you as collaborator. I'd suggest that we develop these code there for the time being, since there is no need to make PRs in the main repo and test it every time we make some slight progress. What's your opinion?

**Comment by matrohon** [comment]
@yf-yang great, I totally agree, let's submit a PR if necessary once the code is mature.

I just discussed with the skydive team. They told me that the simplest approach would be to add a NSM probe in the analyzer. It has a full view of the skydive graph, so the probe would be notified when any element is added into it.

They also gave me a link of a "dumb" probe which we can use as a skeleton for the NSM probe:
https://github.com/safchain/skydive/commit/67eea02074311a1cb720e3833b82a4338b47e652

The NSM probe would use the NSMMonitorClient to connect to every nsmd and to be notified of any event related to  CrossConnects.

**Comment by yf-yang** [comment]
@matrohon Before that, may I ask you some questions about the architecture?
To be specific, it could be split into several questions:
1. Since Skydive is a service, should each node has a pod of SD or in the whole cluster there is only one pod?
2. If there are multiple pods, since there is only one frontend server, where should we put that server? In the cluster or out of the cluster? Which node and which pod? Is there some intermediate entity in charge of gathering information from multiple MonitorClient on multiple pods then pass them to the frontend server?
3. How to set those configurations in that skydive configuration yml?
I don't quite understand the [architecture picture here](http://skydive.network/documentation/#architecture), so when it comes to deploying SD in K8s, I'm always confused. Could you please explain the architecture? Then I'd like to draw a diagram before we start to move the code into SD.
Thanks!

**Comment by matrohon** [comment]
for deploying skydive in k8s you will have 2 main components :

- agents which run on each node in a pod (k8s daemonset): they are hosting some probes to monitor what is happening in terms of networking on each node, and they create the local graph. This graph is periodically sent to an analyzer

- one (or several when deployed in cluster mode) analyzer which run in a pod: it receives every local graph from the agents and build the global graph. It also optionally send this global graph to a storage backend (typically elasticsearch). some probes can be run at the analyzer level, typically probes that fetch datas from controllers (neutron, k8s api...).

For a deployment of skydive in a k8s cluster, you can use k8s configmap to store configuration used by each skydive components, as done in:

https://github.com/ligato/networkservicemesh/blob/2f7acd0c1de706022c431e258fbf13e990956aaa/scripts/vagrant/skydive.yaml#L24

**Comment by edwarnicke** [comment]
@matrohon @yf-yang As of last night, on master, we have working remote cross connects :)

So if you do

make k8s-deploy

you should get a somewhat interesting topology that spans two nodes :)

**Comment by matrohon** [comment]
It is working fine for local links, and @yf-yang as written the skeleton for the remote connections

![sd](https://user-images.githubusercontent.com/1970121/49302309-0b4c0200-f4c7-11e8-8193-1dac41c82046.png)

As you can see in the screenshot, we can highligth nsm crossconnect link by selecting the correct metadata, added by our probe along with the link, in the skydive highlight box

the WIP code is hosted here for now: https://github.com/Orange-OpenSource/skydive/tree/nsm

It still needs a lot of refactoring and hardening, though.

@yf-yang  there was no issue with the inodes number, but our probe needed to listen at skydive graph events, because it was receiving nsm monitoring events before it even knew about inodes.

I also included a skydive patch submitted by @safchain, to mitigate the UI behavior, and allow highlighting of Edges.

**Comment by edwarnicke** [comment]
@matrohon Is it possible to get these visualizations to be a bit less busy so its easier to see the NSM links?

**Comment by yf-yang** [comment]
@edwarnicke I think we are unable to make changes with other nodes since they are provided by other of probes (Or we can do it in a brute way, but it will destruct modules). One possible way to achieve this is to remove some probes so that those nodes are less. Not sure which probes are utilized in the picture, so this may not work if they are already the minimum requirement. The most possible way is to make the nsm links as obtrusive as possible

**Comment by safchain** [comment]
@yf-yang @edwarnicke there is a way to write filter in order to just have a subset of the graph. It is available on the web UI (tiny top left icon). One can write such filter in the config file in order to have it applied by default. I think there is no need to disable something.

**Comment by edwarnicke** [comment]
@safchain @yf-yang @matrohon Could we lead with such a config file?  Its super cool that Skydive can show us the interrelationship of different topologies, but it would be awesome to start focused on the NSM topology :)

**Comment by yf-yang** [comment]
@safchain When I apply a filter on vertexes, the dark grey polygon that shows the node disappears. How to tell the filter to keep some backgrounds?
<img width="312" alt="screen shot 2018-12-01 at 3 37 42 pm" src="https://user-images.githubusercontent.com/36890796/49334018-3e51cc80-f57f-11e8-8c58-50ca8fc4dac5.png">

<img width="605" alt="screen shot 2018-12-01 at 3 36 18 pm" src="https://user-images.githubusercontent.com/36890796/49334005-e7e48e00-f57e-11e8-985e-0669ba790546.png">

**Comment by edwarnicke** [comment]
@yf-yang Looking quite a bit better :)  Question, in addition to getting the node highlighting back, is it possible to filter out the nodes in K8s namespace "kube-system" ?

**Comment by safchain** [comment]
I would use something which is not (yet) documented which allows you to do like an union. I would select netns that you are interrested in and the host nodes. This can be done with something like this.

```
G.V().Has('Type', 'netns', regex('K8S.namespace', '.*aaa.*')).as('mynetns').G.V().Has('Type', 'host').as('myhost').select('mynetns', 'myhost')
```
Basically you do 2 requests, the result is stored in two variables 'mynetns', 'myhost', then you select them as a final result. Of course you can use more than two.

I'm not sure about the "K8S.namespace" key, you need to check with the metadata available on the node.

**Comment by edwarnicke** [comment]
@safchain Cool, you get what we are trying to do wrt the namespace key right?  Suppress the nodes from the kube-system namespace :)

**Comment by matrohon** [comment]
I think this issue is now almost totally fixed by #576 
We now need to submit the probe to skydive, and update docker/Dockerfile.skydive to use 'FROM skydive/skydive'.

**Comment by edwarnicke** [comment]
@matrohon  @yf-yang have a look at #580 

It does two things:

1.  It provides ```make skydive-url``` to output the skydive url
2. Rather than building a new container to get the skydive.yml config file, its now in a config map.

**Comment by edwarnicke** [comment]
It would be nice if in the 'Metatdata' for the link we could have an expandable tree like:

* Id - CrossConnect.Id
* NetworkService - Same as now
* Payload 
* Source
** IP - Connection.ConnectionContext.SrcIP
** MechanismType - Connection.Mechanism.Type
** MechanismParameters - Connection.Mechanism.Parameters as a tree of key value pairs
** Labels - Connection.Labels
** SourceNSM - Connection.SourceNetworkServiceManager (if remote.Connection)
** DestinationNSM - Connection.DestionationNetworkServiceManager (if remote.Connection)
** NetworkServiceEndpoint - Connection.NetworkServiceEndpointName (if remote.Connection)
* Destination
** IP - Connection.ConnectionContext.DstIp
** MechanismType - Connection.Mechanism.Type
** MechanismParameters - Connection.Mechanism.Parameters as a tree of key value pairs
** Labels - Connection.Labels
** SourceNSM - Connection.SourceNetworkServiceManager (if remote.Connection)
** DestinationNSM - Connection.DestionationNetworkServiceManager (if remote.Connection)
** NetworkServiceEndpoint - Connection.NetworkServiceEndpointName (if remote.Connection)

**Comment by yf-yang** [comment]
@matrohon I've made a PR to nsm branch, by the way Ed suggests to make the metadata collapsable and default to collapsed. Now it is only supported for arrays, but not mappings. I guess the source is collapse.js and object-detail.js respectively in static/js/components, but I failed to modify them because I know almost nothing about js. Could you please take care of that?

**Comment by safchain** [comment]
@yf-yang I'll take care of this.

**Comment by edwarnicke** [comment]
@safchain Many thanks :)

**Comment by matrohon** [comment]
@safchain I was thinking having a dedicated collapsable menu in the right panel, like the docker menu. Thoughs?

**Comment by matrohon** [comment]
I just pushed a new image (matrohon/skydive:nsm.0.5) with : 

- arrows;
- more distance between hosts;
- more nsm crossconnect details in the metadatas;

Thanks @safchain and @yf-yang 

@safchain is taking care of the collapsable mappings. We'll see if a dedicated nsm panel is really needed.

**Comment by matrohon** [comment]
it is working great thanks @safchain 

your patch is avalaible in my latest image: matrohon/skydive:nsm.0.6

**Comment by matrohon** [comment]
@yf-yang thanks to your patch, we now have more information on the crossconnect links in the metadatas.
However, since you're using a map, those metadatas are displayed in alphabetical order.

As discussed with @edwarnicke it would be nice to have those metadatas displayed in a strict order: 
CrossConnectId   
NetworkService           
Payload
Source
Destination

As discussed with @safchain, using an array instead of a a map would be the option to force the order of the keys of displayed metadatas.

**Comment by matrohon** [comment]
We also have to connect interfaces instead of  namespaces when the Type is a KERNEL_INTERFACE.

**Comment by edwarnicke** [comment]
In looking at:

![image](https://user-images.githubusercontent.com/1455208/49622883-ff61b380-f990-11e8-9396-1be2c5f52de4.png)

I noticed that Source/Destination panels are only showing the local information for the Source and Destination for remote links.

As far as this goes, this is the correct behavior.

For Remote Links, could we add a 'Via' Panel in between Source and Destination showing the information for the remote.Connection between the two CrossConnects?

**Comment by matrohon** [comment]
@nickolaev I think the first version of the nsm probe for skydive is fine, all the requirements have been developed. Some evolutions may occur in the future, but dedicated issues should target those evolutions.

So to me, this issue can  be closed.
